feat: add documentation for java springboot traces using Otel operator (#1371)
SigNoz/signoz.io: data/docs/instrumentation/opentelemetry-springboot.mdx

</TabItem>
<TabItem value="k8s" label="Kubernetes">

> You can instrument sending traces from Java application using one of the following methods:

Author: ```suggestion
You can auto-instrument sending traces from Java application using one of the following methods:
```

  config: |
    receivers:
      otlp:
        protocols:
          grpc:
            endpoint: 0.0.0.0:4317
          http:
            endpoint: 0.0.0.0:4318
    processors:
      batch: {}
    exporters:
      debug: {}
      otlp:
        endpoint: "ingest.{region}.signoz.cloud:443" # replace {region} with your region
        tls:
          insecure: false
        headers:
          "signoz-ingestion-key": "{signoz-token}" # Obtain from https://{your-signoz-url}/settings/ingestion-settings
    service:
      pipelines:
        traces:
          receivers: [otlp]
          processors: [batch]
          exporters: [otlp]
```

Apply the above yaml file using the following command:

```bash
kubectl apply -f otel-collector.yaml
```

#### Step 4: Setup the Instrumentation instance

Once the OpenTelemetry Collector instance has been deployed, the next step will be to create an instrumentation instance, which will be responsible for sending OTLP data to the OTel Collector.

Create a file `instrumentation.yaml` with the following contents:

```yaml:instrumentation.yaml
apiVersion: opentelemetry.io/v1alpha1
kind: Instrumentation
metadata:
  name: traces-instrumentation
spec:
  exporter:

>     endpoint: http://otel-collector-collector:4317

Author: ```suggestion
  exporter:
    endpoint: https://ingest.{region}.signoz.cloud:443  # replace {region} with your region
  env:
    - name: OTEL_EXPORTER_OTLP_HEADERS
      value: signoz-ingestion-key="{signoz-token}"  # Obtain from https://{your-signoz-url}/settings/ingestion-settings
    - name: OTEL_EXPORTER_OTLP_INSECURE
      value: "false"
```

1. Using Kubernetes OTel Operator
2. Using OTel Collector Agent

<Tabs>
<TabItem value="k8s-otel-operator" label="K8s OTel Operator" default>

For Java application deployed on Kubernetes, you can auto-instrument the traces using Kubernetes OpenTelemetry Operator. 

An [OpenTelemetry Operator](https://opentelemetry.io/docs/kubernetes/operator) is a Kubernetes Operator that manages 
[OpenTelemetry Collectors](https://signoz.io/blog/opentelemetry-collector-complete-guide/) and auto-instrumentation of 
workloads. It basically simplifies the deployment and management of OpenTelemetry in a Kubernetes environment.

The OpenTelemetry Operator provides two Custom Resource Definitions (CRDs):

- `OpenTelemetryCollector`
- `Instrumentation`

The `OpenTelemetryCollector` CRD allows you to deploy and manage OpenTelemetry Collectors in your Kubernetes cluster.

The `Instrumentation` CRD allows you to configure and inject OpenTelemetry auto-instrumentation libraries into your workloads.

Here are the steps you need to follow to auto-instrument Java application using OTel Operator:

#### Step 1: Install cert-manager

Run the following commands to apply cert manager.

```bash

> $ kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.16.1/cert-manager.yaml

Reviewer: ```suggestion
kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.16.1/cert-manager.yaml
```

> $ kubectl wait --for=condition=Available deployments/cert-manager -n cert-manager

Reviewer: ```suggestion
kubectl wait --for=condition=Available deployments/cert-manager -n cert-manager
```

```

#### Step 2: Install OpenTelemetry Operator

To install the operator in the existing K8s cluster, run the following command:

```bash
kubectl apply -f https://github.com/open-telemetry/opentelemetry-operator/releases/download/v0.116.0/opentelemetry-operator.yaml
```

Installing the OpenTelemetry Operator sets up the necessary components and configurations to enable the observability and monitoring of applications running in the cluster.

#### Step 3: Setup the OpenTelemetry Collector instance

Once the `opentelemetry-operator` has been deployed, you can proceed with the creation of the OpenTelemetry Collector (`otelcol`) instance. The OpenTelemetry Collector collects, processes, and exports telemetry data.

There are different deployment modes for the OpenTelemetryCollector, and you can specify them in the spec.mode section of the custom resource. The available deployment modes are:

> - Daemonset

Reviewer: If there is any link available to explain these different modes, we should hyperlink to that. If not, we should give a brief about when to use the different modes.

- Sidecar
- StatefulSet
- Deployment (default mode)

The default method - the `Deployment` mode, will be used here.

To create a simple instance of the OpenTelemetry Collector, create a file `otel-collector.yaml` with the followiung contents:

```yaml:otel-collector.yaml
apiVersion: opentelemetry.io/v1alpha1
kind: OpenTelemetryCollector
metadata:
  name: otel-collector
spec:
  mode: deployment
  config: |
    receivers:
      otlp:
        protocols:
          grpc:
            endpoint: 0.0.0.0:4317
          http:
            endpoint: 0.0.0.0:4318
    processors:
      batch: {}
    exporters:
      debug: {}
      otlp:

>         endpoint: "ingest.{region}.signoz.cloud:443" # replace {region} with your region

Reviewer: ```suggestion
        endpoint: "ingest.<region>.signoz.cloud:443" # replace <region> with your region of SigNoz Cloud
```

        tls:
          insecure: false
        headers:

>           "signoz-ingestion-key": "{signoz-token}" # Obtain from https://{your-signoz-url}/settings/ingestion-settings

Reviewer: ```suggestion
          "signoz-ingestion-key": "<your-ingestion-key>" # Obtain from https://{your-signoz-tenant-url}/settings/ingestion-settings
```

    service:
      pipelines:
        traces:
          receivers: [otlp]
          processors: [batch]
          exporters: [otlp]
```

> 

Reviewer: Mention about `<region>` and `<your-ingestion-key>` in the same way as in the VM section of this doc.

> 

Author: ```suggestion

- Set the `<region>` to match your SigNoz Cloud [region](https://signoz.io/docs/ingestion/signoz-cloud/overview/#endpoint)
- Replace `<your-ingestion-key>` with your SigNoz [ingestion key](https://signoz.io/docs/ingestion/signoz-cloud/keys/)

```

apiVersion: apps/v1
kind: Deployment
metadata:
  name: spring-petclinic
spec:
  selector:
    matchLabels:
      app: spring-petclinic
  replicas: 1
  template:
    metadata:
      labels:
        app: spring-petclinic
      annotations:
        instrumentation.opentelemetry.io/inject-java: "true"
        resource.opentelemetry.io/service.name: "spring-petclinic"
    spec:
      containers:
      - name: app
        image: ghcr.io/pavolloffay/spring-petclinic:latest
        ports:
        - containerPort: 8080
```

Note that we have created the above deployment file for [spring-petclinic](https://github.com/SigNoz/spring-petclinic) Java application.

It is important to add the following annotation under `spec > template > metadata > annotations`:

```
instrumentation.opentelemetry.io/inject-java: "true"`
```

This helps in auto-instrumenting the traces from the Java application.

Apply the deployment using the following command:

```bash
kubectl apply -f deployment.yaml
```

With this, the auto-instrumentation of traces for Java application is ready.

####  Step 6: Running the Java application

In order to run the application on port 8080, run the following commands:

> ```bash

Reviewer: This run command is specific to the spring-petclinic application. Please make it generic. Also keep in mind that application might not run on the local machine.

If the user has the freedom to choose how they run the application, we can skip the run command altogether.

Author: This can be run on the same machine from where the user is executing all the kubernetes commands.

Author: I have made the command generic.

> $ export POD_NAME=$(kubectl get pod -l app=spring-petclinic -o jsonpath="{.items[0].metadata.name}")

Author: ```suggestion
$ export POD_NAME=$(kubectl get pod -l app=<springboot-application> -o jsonpath="{.items[0].metadata.name}")  # springboot application is `spring-petclinic` in this case.   
```

          processors: [batch]
          exporters: [otlp]
```

Apply the above yaml file using the following command:

```bash
kubectl apply -f otel-collector.yaml
```

#### Step 4: Setup the Instrumentation instance

Once the OpenTelemetry Collector instance has been deployed, the next step will be to create an instrumentation instance, which will be responsible for sending OTLP data to the OTel Collector.

Create a file `instrumentation.yaml` with the following contents:

```yaml:instrumentation.yaml
apiVersion: opentelemetry.io/v1alpha1
kind: Instrumentation
metadata:
  name: traces-instrumentation
spec:
  exporter:
  exporter:
    endpoint: https://ingest.{region}.signoz.cloud:443  # replace {region} with your region
  env:
    - name: OTEL_EXPORTER_OTLP_HEADERS
      value: signoz-ingestion-key="{signoz-token}"  # Obtain from https://{your-signoz-url}/settings/ingestion-settings
    - name: OTEL_EXPORTER_OTLP_INSECURE
      value: "false"
  propagators:
    - tracecontext
    - baggage
    - b3
  sampler:
    type: parentbased_traceidratio
    argument: "1"
  java:
    image: ghcr.io/open-telemetry/opentelemetry-operator/autoinstrumentation-java:latest
```

> 

Author: ```suggestion
```

          processors: [batch]
          exporters: [otlp]
```

Apply the above yaml file using the following command:

```bash
kubectl apply -f otel-collector.yaml
```

#### Step 4: Setup the Instrumentation instance

Once the OpenTelemetry Collector instance has been deployed, the next step will be to create an instrumentation instance, which will be responsible for sending OTLP data to the OTel Collector.

Create a file `instrumentation.yaml` with the following contents:

```yaml:instrumentation.yaml
apiVersion: opentelemetry.io/v1alpha1
kind: Instrumentation
metadata:
  name: traces-instrumentation
spec:
  exporter:
  exporter:
    endpoint: https://ingest.{region}.signoz.cloud:443  # replace {region} with your region
  env:
    - name: OTEL_EXPORTER_OTLP_HEADERS
      value: signoz-ingestion-key="{signoz-token}"  # Obtain from https://{your-signoz-url}/settings/ingestion-settings
    - name: OTEL_EXPORTER_OTLP_INSECURE
      value: "false"
  propagators:
    - tracecontext
    - baggage
    - b3
  sampler:
    type: parentbased_traceidratio
    argument: "1"
  java:
    image: ghcr.io/open-telemetry/opentelemetry-operator/autoinstrumentation-java:latest
```

> Note that the ecporter endpoint is based on OpenTelemetryCollector instance name that we have setup in Step 3.

Author: ```suggestion
```

Apply the above instrumentation using the following command:

```bash
kubectl apply -f instrumentation.yaml
```

#### Step 5: Auto-instrument your Java app with OpenTelemetry

Create `deployment.yaml` file for your Java application as follows:

```yaml:deployment.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: spring-petclinic
spec:
  selector:
    matchLabels:
      app: spring-petclinic
  replicas: 1
  template:
    metadata:
      labels:
        app: spring-petclinic
      annotations:
        instrumentation.opentelemetry.io/inject-java: "true"
        resource.opentelemetry.io/service.name: "spring-petclinic"
    spec:
      containers:
      - name: app
        image: ghcr.io/pavolloffay/spring-petclinic:latest
        ports:
        - containerPort: 8080
```

> Note that we have created the above deployment file for spring-petclinic Java application.

Author: ```suggestion
Note that we have created the above deployment file for [spring-petclinic](https://github.com/SigNoz/spring-petclinic) Java application.
```

</TabItem>
<TabItem value="k8s" label="Kubernetes">

> You can auto-instrument sending traces from Java application using one of the following methods:

Author: ```suggestion
You can auto-instrument sending traces from Java Springboot application using one of the following methods:
```

1. Using Kubernetes OTel Operator
2. Using OTel Collector Agent

<Tabs>
<TabItem value="k8s-otel-operator" label="K8s OTel Operator" default>

> For Java application deployed on Kubernetes, you can auto-instrument the traces using Kubernetes OpenTelemetry Operator. 

Author: ```suggestion
For Java Springboot application deployed on Kubernetes, you can auto-instrument the traces using Kubernetes OpenTelemetry Operator. 
```

An [OpenTelemetry Operator](https://opentelemetry.io/docs/kubernetes/operator) is a Kubernetes Operator that manages 
[OpenTelemetry Collectors](https://signoz.io/blog/opentelemetry-collector-complete-guide/) and auto-instrumentation of 
workloads. It basically simplifies the deployment and management of OpenTelemetry in a Kubernetes environment.

The OpenTelemetry Operator provides two Custom Resource Definitions (CRDs):

- `OpenTelemetryCollector`
- `Instrumentation`

The `OpenTelemetryCollector` CRD allows you to deploy and manage OpenTelemetry Collectors in your Kubernetes cluster.

The `Instrumentation` CRD allows you to configure and inject OpenTelemetry auto-instrumentation libraries into your workloads.

> Here are the steps you need to follow to auto-instrument Java application using OTel Operator:

Author: ```suggestion
Here are the steps you need to follow to auto-instrument Springboot application using OTel Operator:
```

#### Step 1: Install cert-manager

Run the following commands to apply cert manager.

```bash
kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.16.1/cert-manager.yaml

kubectl wait --for=condition=Available deployments/cert-manager -n cert-manager
```

#### Step 2: Install OpenTelemetry Operator

To install the operator in the existing K8s cluster, run the following command:

```bash
kubectl apply -f https://github.com/open-telemetry/opentelemetry-operator/releases/download/v0.116.0/opentelemetry-operator.yaml
```

Installing the OpenTelemetry Operator sets up the necessary components and configurations to enable the observability and monitoring of applications running in the cluster.

#### Step 3: Setup the OpenTelemetry Collector instance

Once the `opentelemetry-operator` has been deployed, you can proceed with the creation of the OpenTelemetry Collector (`otelcol`) instance. The OpenTelemetry Collector collects, processes, and exports telemetry data.

There are different deployment modes for the OpenTelemetryCollector, and you can specify them in the spec.mode section of the custom resource. The available deployment modes are:

> - Daemonset

Author: ```suggestion
- [Daemonset](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/)
```

> - Sidecar

Author: ```suggestion
- [Sidecar](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/)
```

> - StatefulSet

Author: ```suggestion
- [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/)
```

> - Deployment (default mode)

Author: ```suggestion
- [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) (default mode)
```

```bash
kubectl apply -f otel-collector.yaml
```

#### Step 4: Setup the Instrumentation instance

Once the OpenTelemetry Collector instance has been deployed, the next step will be to create an instrumentation instance, which will be responsible for sending OTLP data to the OTel Collector.

Create a file `instrumentation.yaml` with the following contents:

```yaml:instrumentation.yaml
apiVersion: opentelemetry.io/v1alpha1
kind: Instrumentation
metadata:
  name: traces-instrumentation
spec:
  exporter:
  exporter:
    endpoint: https://ingest.{region}.signoz.cloud:443  # replace {region} with your region
  env:
    - name: OTEL_EXPORTER_OTLP_HEADERS
      value: signoz-ingestion-key="{signoz-token}"  # Obtain from https://{your-signoz-url}/settings/ingestion-settings
    - name: OTEL_EXPORTER_OTLP_INSECURE
      value: "false"
  propagators:
    - tracecontext
    - baggage
    - b3
  sampler:
    type: parentbased_traceidratio
    argument: "1"
  java:
    image: ghcr.io/open-telemetry/opentelemetry-operator/autoinstrumentation-java:latest
```

Apply the above instrumentation using the following command:

```bash
kubectl apply -f instrumentation.yaml
```

> #### Step 5: Auto-instrument your Java app with OpenTelemetry

Author: ```suggestion
#### Step 5: Auto-instrument your Springboot app with OpenTelemetry
```

> Create `deployment.yaml` file for your Java application as follows:

Author: ```suggestion
Create `deployment.yaml` file for your Springboot application as follows:
```

```yaml:deployment.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: spring-petclinic
spec:
  selector:
    matchLabels:
      app: spring-petclinic
  replicas: 1
  template:
    metadata:
      labels:
        app: spring-petclinic
      annotations:
        instrumentation.opentelemetry.io/inject-java: "true"
        resource.opentelemetry.io/service.name: "spring-petclinic"
    spec:
      containers:
      - name: app
        image: ghcr.io/pavolloffay/spring-petclinic:latest
        ports:
        - containerPort: 8080
```

> Note that we have created the above deployment file for [spring-petclinic](https://github.com/SigNoz/spring-petclinic) Java application.

Author: ```suggestion
Note that we have created the above deployment file for [spring-petclinic](https://github.com/SigNoz/spring-petclinic) Java Springboot application.
```

It is important to add the following annotation under `spec > template > metadata > annotations`:

```
instrumentation.opentelemetry.io/inject-java: "true"`
```

This helps in auto-instrumenting the traces from the Java application.

Apply the deployment using the following command:

```bash
kubectl apply -f deployment.yaml
```

> With this, the auto-instrumentation of traces for Java application is ready.

Author: ```suggestion
With this, the auto-instrumentation of traces for Java Springboot application is ready.
```

####  Step 6: Running the Java application

In order to run the application on port 8080, run the following commands:

```bash

> $ export POD_NAME=$(kubectl get pod -l app=<springboot-application> -o jsonpath="{.items[0].metadata.name}")  # springboot application is `spring-petclinic` in this case.   

Author: ```suggestion
$ export POD_NAME=$(kubectl get pod -l app=<service-name> -o jsonpath="{.items[0].metadata.name}")  # service name is `spring-petclinic` in this case.   
```

> ####  Step 6: Running the Java application

Author: ```suggestion
####  Step 6: Running the Springboot application
```

$ kubectl port-forward ${POD_NAME} 8080:8080
```

The application will start running on port 8080 on your local machine. You can access the application by entering 
the following URL on your browser: http://localhost:8080

You can validate if your application is sending traces to SigNoz cloud by following the instructions [here](#validating-instrumentation-by-checking-for-traces).

</TabItem>
<TabItem value="otel-collector-agent" label="OTel Collector Agent">

> For Java application deployed on Kubernetes, you need to install OTel Collector agent in your k8s infra to collect and send traces to SigNoz Cloud. You can find the

Author: ```suggestion
For Java Springboot application deployed on Kubernetes, you need to install OTel Collector agent in your k8s infra to collect and send traces to SigNoz Cloud. You can find the
```

```bash
kubectl apply -f https://github.com/open-telemetry/opentelemetry-operator/releases/download/v0.116.0/opentelemetry-operator.yaml
```

Installing the OpenTelemetry Operator sets up the necessary components and configurations to enable the observability and monitoring of applications running in the cluster.

#### Step 3: Setup the OpenTelemetry Collector instance

Once the `opentelemetry-operator` has been deployed, you can proceed with the creation of the OpenTelemetry Collector (`otelcol`) instance. The OpenTelemetry Collector collects, processes, and exports telemetry data.

There are different deployment modes for the OpenTelemetryCollector, and you can specify them in the spec.mode section of the custom resource. The available deployment modes are:

- [Daemonset](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/)
- [Sidecar](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/)
- [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/)
- [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) (default mode)

The default method - the `Deployment` mode, will be used here.

> To create a simple instance of the OpenTelemetry Collector, create a file `otel-collector.yaml` with the followiung contents:

Author: ```suggestion
To create a simple instance of the OpenTelemetry Collector, create a file `otel-collector.yaml` with the following contents:
```

You can validate if your application is sending traces to SigNoz cloud by following the instructions [here](#validating-instrumentation-by-checking-for-traces).

In case you encounter an issue where all applications do not get listed in the services section then please refer to the [troubleshooting section](#troubleshooting-your-installation).

> </TabItem>

Author: ```suggestion
</TabItem>
</Tabs>
```
